### DISCUSSION TITLE
Fix for "No tap interface is available, cannot proceed."

### DISCUSSION THREAD

**Discussion by JohnnyZig** [pgj/freebsd-wifibox#130] [Q&A]
I was	 trying to use wifibox on 15-CURRENT.
Below are my problem and fix documented in hope that it will help someone.

HW
```
pciconf -lv | grep ppt0 -A3
ppt0@pci0:4:0:0:        class=0x028000 rev=0x88 hdr=0x00 vendor=0x8086 device=0x24fd subvendor=0x8086 subdevice=0x1130
    vendor     = 'Intel Corporation'
    device     = 'Wireless 8265 / 8275'
    class      = network
```

Kernel modules are loaded
```
kldstat | grep -E "vmm|linux"
 3    5 0xffffffff81aea000    20af8 linux_common.ko
 7    1 0xffffffff82c00000   33f438 vmm.ko
11    3 0xffffffff83177000     3080 linuxkpi_hdmi.ko
24    1 0xffffffff831d3000    31290 linux.ko
26    1 0xffffffff8320c000    2e3b8 linux64.ko
```
Kernel modules iwm and iwlwifi are not loaded:
```
kldstat -v | grep -i if_
	198 if_lo
	202 if_vlan
	201 if_tap
	200 if_tun
	199 if_tuntap
	197 if_gif
```

/etc/rc.conf
```
devmatch_blocklist="if_iwm if_iwlwifi"
kld_list="${kld_list} vmm"
wifibox_enable="YES"
ifconfig_wifibox0="SYNCDHCP"
background_dhclient_wifibox0="YES"
defaultroute_delay="0"
```

Default configuration files in /usr/local/etc/wifibox are used (pci0:4:0:0 is already default in bhyve.conf).

I am unable to start wifibox service
```
# service wifibox start
Starting wifibox.....ERROR: No tap interface is available, cannot proceed.
ERROR: Guest could not be started.

# service netif start wifibox0
ifconfig: interface wifibox0 does not exist
Starting dhclient.
ifconfig: interface wifibox0 does not exist
wifibox0: not found
exiting.
/etc/rc.d/dhclient: WARNING: failed to start dhclient
Starting Network: wifibox0.
ifconfig: interface wifibox0 does not exist
```

I've noticed that after every "service wifibox start" another tap interface is created:
```
lo0: flags=1008049<UP,LOOPBACK,RUNNING,MULTICAST,LOWER_UP> metric 0 mtu 16384
        options=680003<RXCSUM,TXCSUM,LINKSTATE,RXCSUM_IPV6,TXCSUM_IPV6>
        inet 127.0.0.1 netmask 0xff000000
        inet6 ::1 prefixlen 128
        inet6 fe80::1%lo0 prefixlen 64 scopeid 0x1
        groups: lo
        nd6 options=21<PERFORMNUD,AUTO_LINKLOCAL>
tap0: flags=8802<BROADCAST,SIMPLEX,MULTICAST> metric 0 mtu 1500
        options=80000<LINKSTATE>
        ether 58:9c:fc:10:ff:fe
        groups: tap
        media: Ethernet 1000baseT <full-duplex>
        status: no carrier
        nd6 options=29<PERFORMNUD,IFDISABLED,AUTO_LINKLOCAL>
tap1: flags=8802<BROADCAST,SIMPLEX,MULTICAST> metric 0 mtu 1500
        options=80000<LINKSTATE>
        ether 58:9c:fc:10:f4:2e
        groups: tap
        media: Ethernet 1000baseT <full-duplex>
        status: no carrier
        nd6 options=29<PERFORMNUD,IFDISABLED,AUTO_LINKLOCAL>
...
tapN: ...
```

/var/log/wifibox.log

```
2024-09-28T11:15:59+0200 ERROR No tap interface is available, cannot proceed
2024-09-28T11:16:00+0200 ERROR Guest could not be started
```

I was using custom kernel config.
Fix for the error is to build & load if_bridge kmod:
```
kldstat | grep -i if_bridge
28    1 0xffffffff8323e000     78d8 if_bridge.k
```
Now wifibox starts without errors :smiley:

**Comment by pgj** [comment]
Thanks for documenting this.  Wifibox heavily relies on the functionality provided by the `if_bridge` kernel module and since it is shipped with the `GENERIC` kernel, its presence is not explicitly checked.  Hence I would say custom kernel configurations are not supported, especially in combination with `CURRENT`, where technically your mileage may vary all the time.  But that is good to be aware of this implicit dependency at least.

**Reply by mord0d** [reply]
Why do we need a bridge to be hardcoded, especially with hostap?

I've faced the same error on FreeBSD 15.0-RELEASE-p4 (with GENERIC kernel): wifibox can't see the interface it just created. Moreover, it can't recognise its own bridge named `wifibox0` and creates new `bridgeN` each time.

**Reply by pgj** [reply]
The primary reason for hardcoding the interface name is that Wifibox cannot (and should not) work with multiple interfaces as that is a non-supported, albeit theoretically possible scenario.  Personally, I do not own a device where there are multiple wireless cards present, therefore I cannot test that.

**Reply by mord0d** [reply]
I mean, why wifibox creates the bridge? What if we already have a bridge (I have 8 separate Ethernet cards on my router so I already have a bridge)?

I've read the code, you use the bridge to determine if a tap interface is for Wifibox. You can use `ifconfig tap create group wifibox up` instead, and match it by the group.

I don't want to touch the code yet, because I'm new to the project and don't know the use-cases and can break things for people easily.

**Reply by pgj** [reply]
You are right, the bridge is not a strict requirement for making it work.  The bridge is created to associate a stable, intuitive name with the interface that Wifibox uses.  It can help with simplifying the documentation and easier for the users to work that.

Note that even though `ifconfig` supports renaming the `tap` interface or even create that with a specific name, `bhyve` simply will not start the virtual machine.  According to my observations, `if_tuntap` will eventually create a `/dev/tapN` device and the aliased name will just be a symbolic link pointing to that, with which `bhyve` has problems on opening.  Perhaps it is just a bug, but that is what I could find on the subject so far.

**Reply by mord0d** [reply]
Yes, we cannot rename a tap if we use it with bhyve, but we can assign it to a custom group ([churchers/vm-bhyve](https://github.com/churchers/vm-bhyve) does this) and/or add a description.

We create a tap interface (it could be wrapped in a subshell to get the `tapN`):

```sh
ifconfig tap create group wifibox up
```

Then we match if it's ours:

```sh
for i in $(ifconfig -l); do
  case "${i}" in
    (tap*)
      ifconfig "${i}" | grep '^[%s]*groups .*wifibox' \
        && return 0 # we found our tap
      # will continue if a tap not matched
      ;;
    (*) ;; # skip non-tap and renamed tap's
  esac
done
return 1 # we didn't found our tap
```

**Reply by pgj** [reply]
Detecting if the interface has already been assigned to Wifibox is the least problem.  That is more rooted in establishing a user experience and abstracting away from the implementation details.  (Why users should know that the network traffic is coming from a VM via a tap interface?)   Let me also note that I have been using `vm-bhyve` (well, that inspired the command-line interface of Wifibox) with a "switch" (`vm-public`), which is a bridge too.

**Reply by mord0d** [reply]
> Detecting if the interface has already been assigned to Wifibox is the least problem.  That is more rooted in establishing a user experience and abstracting away from the implementation details.  (Why users should know that the network traffic is coming from a VM via a tap interface?)   Let me also note that I have been using `vm-bhyve` (well, that inspired the command-line interface of Wifibox) with a "switch" (`vm-public`), which is a bridge too.

I agree with you. There's no reason to change `wpa_supplicant`.

But the average user don't need `hostapd`. Otherwise they will learn a lot of things one way or another. And the admin knows what they're doing. They could use their own bridge Link Aggregation, VLAN/VxLAN, tun/gif, forward packets using the chosen firewall *in any possible combination*.

My proposal is not to add support for every possible option, but to simplify and let user make a choice. We could make an option to skip bridge creation, *leaving the default as it is.*

I'm experimenting on my laptop and when I'm sure I won't break anything, I'll file a PR.

*// By the way, vm-bhyve can use admin-defined switch and will never alter its configuration. Though will fail to start if not pre-configured.*

**Comment by Defenso-QTH** [comment]
I am facing the same error even with `if_bridge` loaded on HardenedBSD 14-STABLE.

**Reply by Defenso-QTH** [reply]
I just put a `true` here and it seems to boot fine: https://github.com/pgj/freebsd-wifibox/blob/24c44b26e485d718584160577c06e663f95f2d2b/sbin/wifibox#L299

Note that I am still stuck because of #137

**Reply by mord0d** [reply]
`grub-bhyhe` cannot run with `kern.elf64.allow_wx=0`.

I think we need wifibox-alpine with UEFI loader, since Edk2 works even if `W^X` is disabled.

**Reply by pgj** [reply]
That sounds like an unrelated problem, but it is a good suggestion.  Would you mind creating a ticket about that?

**Reply by mord0d** [reply]
> That sounds like an unrelated problem, but it is a good suggestion.  Would you mind creating a ticket about that?

@pgj,

It is related, because Wifibox uses only `grub-bhyve` to boot the guest. And related to wifibox-alpine, because its linux config doesn't have efi_stub enabled. But it's complicated thing, I have to figure it out by myself before suggesting anything.

Before I will create an issue, I want to know why you decided to use grub2 instead of UEFI and enable `efi_stub` in your Linux kernel config. Maybe it's better to create a separate fork (for both wifibox and wifibox-alpine), because we can't easily switch between boot loaders.

As I said in anothet thread, I don't want to break things.

**Reply by pgj** [reply]
To be honest, I do not know UEFI well enough, and I am open to suggestions (and learn!) about that.  I started with GRUB because it was easier for me.  I do not think it would require a fork, we could just switch over to using UEFI and that is it.  The management of the virtual machine is hidden from the users, they should not make any assumptions (or even know) about whether the VM is booted via GRUB or UEFI.

**Reply by mord0d** [reply]
- Pros of the `grub-bhyve`:
  - It allows to boot a VM via GRUB2.
- Cons of the `grub-bhyve`:
  - It requires `W^X` memory (which is disabled, `kern.elf64.allow_wx=0` and `kern.elf32.allow_wx=0`, on hardened systems);
  - It requires additional *separate* steps (run `grub-bhyve` before `bhyve`) to boot a guest;
  - It can boot only Linux (and other OSes via `chainload`, but it requires additional setup);
  - It cannot do PXE boot (FIXME: I could be wrong about that).

- Pros of the `edk2-bhyve`:
  - It allows to boot almost any (except those that uses `u-boot`) OS;
  - It does not require any additional *separate* steps (an argument to `bhyve` will suffice);
  - It supports PXE boot;
  - It can even boot GRUB2 (`grub2-efi`).
- Cons of the `edk2-bhyve`:
  - It requires UEFI support from the boot loader.

First of all, if we use `edk2-bhyve`, we could throw away GRUB2 from the guest, but we should compile the Linux kernel with `EFI_STUB` (it will grow in size for about 2M, if I recall correctly) option (I can't remember the exact `CONFIG_*` option, but you could easily find it in `make nconfig`).

Second, we don't wait for `grub-bhyve` to preload, just run `bhyve` with `l bootrom,/usr/local/share/uefi-firmware/BHYVE_UEFI.fd`, that's it.

In summary, switching from `grub-bhyve` to `edk2-bhyve` requires changing guest boot-up process in `wifibox` *and* changing the quest (`wifibox-alpine`) itself.

I didn't figure out how to compile `wifibox-alpine` myself yet, facing a lot of errors at `make`, but it's another story.

**Reply by pgj** [reply]
Could you please move this whole comment above to a ticket or at least a separate discussion thread?  I would love to follow up on your comment but I am worried that the whole topic will just get lost here if we keep talking about that in a sub-thread here.

**Comment by mord0d** [comment]
I have a solution for the brifge+tap issue:

```diff
--- a/sbin/wifibox
+++ b/sbin/wifibox
@@ -296,7 +296,7 @@ load_bhyve_conf_values() {
 }
 
 has_bridge_interface() {
-    ${IFCONFIG} | ${GREP} -Fq "${WIFIBOX_IF}: "
+    ${IFCONFIG} -l | ${GREP} -Fq "${WIFIBOX_IF}"
 }
 
 get_tap_interface() {
```

This is easier and doesn't block if some interface gathering its configuration right now.

@pgj, take a look, please.

**Reply by mord0d** [reply]
Or even better:

```sh
has_bridge_interface() {
  local i
  for i in $(${IFCONFIG} -l); do
    case "${i}" in
      ("${WIFIBOX_IF}") return 0 ;;
      (*) ;;
    esac
  done
  return 1
}
```

This is even cheaper than using `grep` and never hit false-positive on interfaces *containing* `wifibox0` in their names as in my previous solution.

**Reply by pgj** [reply]
Yeah, this one looks like a neat simplification.  But I wonder why the original code does not work: it should be the same semantically.

**Reply by mord0d** [reply]
> Yeah, this one looks like a neat simplification.  But I wonder why the original code does not work: it should be the same semantically.

In fact, I can't reproduce the case outside the wifibox script. I have no idea why it doesn't work **for some people** (including me).
